### ISSUE TITLE
How to update height of segmented Button

### ISSUE THREAD

**Issue by darkwood-studios** [TheJuki/KFormMaster#205]
Hi there,

I want to update the height of a segmented button on value change.

This is my approch:

```
segmented<SquareMeterListItem>(1) {
    valueObservers.add { newValue, element ->
        (element as FormSegmentedElement).radioButtonHeight = newValue?.getRadioButtonHeight()
    }
}
```

The method "getRadioButtonHeight" return null or 60

Problem: The app crashes. It looks like a loop.

Any ideas how to change the height of the segmented button on value change?

**Comment by TheJuki** [comment]
valueObservers.add { newValue, element ->
                    options?.forEach {
                        it.height = 50
                        if (it == newValue) {
                            it.height = 70
                        }
                    }
    
                    reInitGroup()
                }

If this is what you mean, this is what I do in the Full Screen Example which works. The selected button's height gets increased while the rest reset their height.

**Comment by darkwood-studios** [comment]
Hi @TheJuki 

this is not working. "it.height = 70" calls SegmentedDrawable.setHeight() in my example. This method is not implemented.

I think "radioButtonHeight" is the correct param. But it calls reInitGroup() which results in al infinite loop. reInitGroup will change the value and the valueObserver will call reInitGroup. - I think it should be possible to change the radioButtonHeight without the call of reinitGroup

**Comment by darkwood-studios** [comment]
My problem is that I have two different optionLists of SegmentedDrawable. One List has a drawable with an name (label) and one list has a drawable without name. If I change the list, the height of the options without a name still have the same height as the list with the name. See my problem on the screenshot:

Maybe you have another idea to solve this.
<img width="494" alt="Bildschirmfoto 2020-08-19 um 09 32 17" src="https://user-images.githubusercontent.com/8148815/90605629-e9e0fb80-e1fe-11ea-8a1d-2ecfab2cd31b.png">
<img width="491" alt="Bildschirmfoto 2020-08-19 um 09 32 23" src="https://user-images.githubusercontent.com/8148815/90605633-ebaabf00-e1fe-11ea-901f-b80cb5f0ba80.png">

**Comment by TheJuki** [comment]
I cannot replicate the infinite loop issue. I tried setting up based on how you described it but valueObservers is only called once. Setting radioButtonHeight is not changing the model value again. Can you provide more code?

**Comment by darkwood-studios** [comment]
Yes I can. 



`/**
* list of SquareMeterListItems
**/
val listItems = mapOf(
                "G" to mapOf(
                    "H" to listOf(
                        SquareMeterListItem(id = 1, name = "30 m²", usage = "6000", drawableRes = R.drawable.ic_home),
                        SquareMeterListItem(id = 2, name = "50 m²", usage = "10000", drawableRes = R.drawable.ic_home),
                        SquareMeterListItem(id = 3, name = "100 m²", usage = "20000", drawableRes = R.drawable.ic_home)
                    ),
                    "G" to listOf(
                        SquareMeterListItem(id = 1, name = null, usage = "24000", drawableRes = R.drawable.ic_building),
                        SquareMeterListItem(id = 2, name = null, usage = "40000", drawableRes = R.drawable.ic_building_2),
                        SquareMeterListItem(id = 3, name = null, usage = "80000", drawableRes = R.drawable.ic_building_3)
                    )
                )
            )

private val squareMeterList = listItems.getValue("G").getValue("H")

/**
* This method changes the sqare meter ist in segmented button
* example: changeUsageButtons("G", "H")
**/
private fun changeUsageButtons(product: String, type: String) {
       val squareMeterListNew = listItems.getValue(product).getValue(type)
       formBuilder.getFormElement<FormSegmentedElement<SquareMeterListItem>>(FormField.TYPE_USAGE_BUTTON_LIST).options = squareMeterListNew

       var selectedItem: Int? = formBuilder.getFormElement<FormSegmentedElement<SquareMeterListItem>>(FormField.TYPE_USAGE_BUTTON_LIST).value?.id?.toInt()
       if (selectedItem == null) {
           selectedItem = 1 // select second one if nothing selected
       } else {
           selectedItem -= 1 // array correction
       }

       formBuilder.getFormElement<FormSegmentedElement<SquareMeterListItem>>(FormField.TYPE_USAGE_BUTTON_LIST).setValue(squareMeterListNew[selectedItem])
}

/**
* Form Elements
**/
number(FormField.TYPE_USAGE_INPUT) {
                title = "Usage"
                value = "2500"
                required = true
                numbersOnly = true
                layoutPaddingBottom = 0
                margins = FormElementMargins(16, 16, 16, 0)
                clearOnFocus = true
}
segmented<SquareMeterListItem>(FormField.TYPE_USAGE_BUTTON_LIST) {
                options = squareMeterList
                value = squareMeterList[1]
                editViewGravity = Gravity.START
                fillSpace = true
                displayDivider = false
                margins = FormElementMargins(16, 0, 16, 0)
                valueObservers.add { newValue, element ->
                    formBuilder.getFormElement<FormNumberEditTextElement>(FormField.TYPE_USAGE_INPUT).setValue(newValue?.usage)
          
                    //(element as FormSegmentedElement).radioButtonHeight = newValue?.getRadioButtonHeight() // todo produces crash
                }
                //radioButtonHeight = this.value?.getRadioButtonHeight() // todo does nothing
}


/**
* Class SquareMeterListItem
**/
data class SquareMeterListItem(val id: Long? = null, val name: String? = null, val usage: String? = "", @DrawableRes override var drawableRes: Int? = 0) : SegmentedDrawable {
    override fun toString(): String {
        return name.orEmpty()
    }

    fun getRadioButtonHeight(): Int? {
        // less height if we don't have a name/label
        if(name == null) {
            return 55
        }

        return null
    }
}
`

**Comment by TheJuki** [comment]
I was able to replicate. displayNewValue is calling checkChild over and over. It does that because if you set the value yourself later it adds a button for some reason. So, I can remove that code, but to set a value without issues you would need to set holdup:

val test = formBuilder.getFormElement<FormSegmentedElement<SquareMeterListItem>>(FormField.TYPE_USAGE_BUTTON_LIST.ordinal)
            test.editView?.let {
                if (it is SegmentedGroup) {
                    it.holdup = true
                }
            }
            test.value = squareMeterList[1]

**Comment by darkwood-studios** [comment]
@TheJuki 
Yes this should work. Is the holdup property still available in 7.0 or will there be a new release?

**Comment by TheJuki** [comment]
I just released [8.0.0](https://github.com/TheJuki/KFormMaster/releases/latest). Please update your dependency and let me know if this issue is resolved for you.

**Comment by TheJuki** [comment]
I am closing this issue. If you have any issues feel free to reopen.
